### ISSUE TITLE
Add documentation for each sniff

### ISSUE THREAD

**Issue by jrfnl** [WordPress-Coding-Standards/WordPress-Coding-Standards#1722]
All sniffs should be accompanied by documentation in the form of a `SniffNameStandard.xml` file.
A `SniffNameStandard.xml` file contains a (short) description of each rule(s) the sniff checks and a code comparison for each check.

Examples can be found in the PHP_CodeSniffer repository:
* https://github.com/squizlabs/PHP_CodeSniffer/blob/master/src/Standards/Generic/Docs/Formatting/MultipleStatementAlignmentStandard.xml
* https://github.com/squizlabs/PHP_CodeSniffer/tree/master/src/Standards/Generic/Docs

The sniff documentation for the complete standard can be shown using the following command:
```bash
vendor/bin/phpcs --standard=WordPress --generator=text
```

The sniff documentation for an individual sniff can be shown using the following command:
```bash
vendor/bin/phpcs --standard=WordPress --generator=text --sniffs=WordPress.Category.SniffName
```

Aside from the `text` generator, there are also generators available for HTML and Markdown, which will allow us to use these documentation files to auto-generate documentation about the sniffs to be placed in, for instance, the wiki.

At this time, none of the WPCS native sniffs have this kind of documentation available.

This issue is intended to track progress for adding this documentation.

**Adding this documentation is a focus-task for the WCEU contributor day 2019.**
During the WCEU Contributor day Gary (@GaryJones ) and Juliette (@jrfnl) will be available to help contributors get set up and answer any questions they may have.

### Process
* When you start working on the documentation for one of these sniffs, please mention it in the WP Slack `#core-coding-standards` channel and mention your GitHub nickname (during the WCEU contributor day) or leave a comment in this issue (after WCEU Contributor day).
* Gary or Juliette will update this issue to show you have claimed that sniff.

#### Preparation
* Create a [GitHub user account](https://www.wikihow.com/Create-an-Account-on-GitHub) (if you haven't got one already).
* [Fork this repository on GitHub](https://help.github.com/en/articles/fork-a-repo).
* Make sure you have both [git](https://git-scm.com/downloads) as well as [Composer](https://getcomposer.org/) installed on your machine.
* [Clone your fork to your local machine](https://help.github.com/en/articles/cloning-a-repository) using your preferred git client.
* Run `composer install` from the command line in the root directory of your local clone of this repo.

#### For each sniff:
* Create a new branch off `develop` and name it `docs/sniffname`.
    ```bash
    git checkout -b "docs/sniffname" "develop"
    ```
* Create the XML file and place it in the correct `WordPress/Docs/Category/` directory.
    You may need to create the directory if it doesn't yet exist.
* To verify what the sniff is checking for, look for `addWarning()` and/or `addError()` function calls in the sniff file in the `WordPress/Sniffs/Category/` directory to see the error messages the sniff generates.
* To get inspiration for code examples, open the sniff's `WordPress/Tests/Category/SniffNameUnitTest.inc` file.
    Correct code will generally be marked with `// OK.`, incorrect code will be marked with `// Bad` or `// Error`.
    If the sniff generates several error messages and you are unsure which error applies to which code, you can run the following command to get a better understanding of which code triggers which error/warning:
    ```bash
    vendor/bin/phpcs -s ./WordPress/Tests/Category/SniffNameUnitTest.inc --standard=WordPress --sniffs=WordPress.Category.SniffName
    ```
* Guidelines for the code samples:
    - Keep it as simple as possible.
    - Don't use code which can be traced back to a specific project.
    - Each line within the code sample should be max 48 characters.
* Once you've created the documentation, test it by running:
    ```bash
    vendor/bin/phpcs --standard=WordPress --generator=text --sniffs=WordPress.Category.SniffName
    ```
* If all is OK, [push the branch upstream](https://help.github.com/en/articles/pushing-to-a-remote) to your fork and [create a pull request](https://help.github.com/en/articles/creating-a-pull-request-from-a-fork) in this repo.
    Please mention this issue in your pull request description `Related to #xxx` so your pull request will show up in this issue.
* Gary and/or Juliette will review your PR and either approve and merge it or leave feedback on the pull request.
* Once a PR has been merged, this issue should be updated and the sniff action item moved to the "Finished" list and marked as complete.

## To Do
- [ ] WordPress.Arrays.ArrayDeclarationSpacing
- [ ] WordPress.Arrays.ArrayIndentation                                   
- [ ] WordPress.Arrays.ArrayKeySpacingRestrictions                        
- [ ] WordPress.Arrays.CommaAfterArrayItem                                
- [ ] WordPress.Arrays.MultipleStatementAlignment                         
- [ ] WordPress.Classes.ClassInstantiation                                
- [ ] WordPress.CodeAnalysis.AssignmentInCondition                        
- [ ] WordPress.CodeAnalysis.EmptyStatement                               
- [ ] WordPress.DB.DirectDatabaseQuery
- [ ] WordPress.DB.PreparedSQL
- [ ] WordPress.DB.PreparedSQLPlaceholders                             
- [ ] WordPress.DB.RestrictedClasses                                   
- [ ] WordPress.DB.RestrictedFunctions                                 
- [ ] WordPress.DB.SlowDBQuery                                         
- [ ] WordPress.Files.FileName                                         
- [ ] WordPress.NamingConventions.PrefixAllGlobals                     
- [ ] WordPress.NamingConventions.ValidFunctionName                    
- [ ] WordPress.NamingConventions.ValidHookName                        
- [ ] WordPress.NamingConventions.ValidVariableName                    
- [ ] WordPress.PHP.DevelopmentFunctions                               
- [ ] WordPress.PHP.DiscouragedPHPFunctions                            
- [ ] WordPress.PHP.DontExtract                                        
- [ ] WordPress.PHP.IniSet                                             
- [ ] WordPress.PHP.NoSilencedErrors                                   
- [ ] WordPress.PHP.POSIXFunctions
- [ ] WordPress.PHP.PregQuoteDelimiter                                 
- [ ] WordPress.PHP.RestrictedPHPFunctions                             
- [ ] WordPress.PHP.StrictComparisons                                  
- [ ] WordPress.PHP.StrictInArray                                      
- [ ] WordPress.PHP.TypeCasts                                          
- [ ] WordPress.PHP.YodaConditions                                     
- [ ] WordPress.Security.EscapeOutput                                  
- [ ] WordPress.Security.NonceVerification                             
- [ ] WordPress.Security.PluginMenuSlug                                
- [ ] WordPress.Security.SafeRedirect                                  
- [ ] WordPress.Security.ValidatedSanitizedInput                       
- [ ] WordPress.Utils.I18nTextDomainFixer                              
- [ ] WordPress.WP.AlternativeFunctions                                
- [ ] WordPress.WP.CapitalPDangit                                      
- [ ] WordPress.WP.CronInterval                                        
- [ ] WordPress.WP.DeprecatedClasses                                   
- [ ] WordPress.WP.DeprecatedFunctions                                 
- [ ] WordPress.WP.DeprecatedParameterValues                           
- [ ] WordPress.WP.DeprecatedParameters                                
- [ ] WordPress.WP.DiscouragedConstants                                
- [ ] WordPress.WP.DiscouragedFunctions                                
- [ ] WordPress.WP.EnqueuedResourceParameters                          
- [ ] WordPress.WP.EnqueuedResources                                   
- [ ] WordPress.WP.GlobalVariablesOverride                             
- [ ] WordPress.WP.I18n                                                
- [ ] WordPress.WP.PostsPerPage                                        
- [ ] WordPress.WP.TimezoneChange                                      
- [ ] WordPress.WhiteSpace.CastStructureSpacing                        
- [ ] WordPress.WhiteSpace.ControlStructureSpacing                     
- [ ] WordPress.WhiteSpace.DisallowInlineTabs                          
- [ ] WordPress.WhiteSpace.OperatorSpacing                             
- [ ] WordPress.WhiteSpace.PrecisionAlignment

## Claimed
None so far

## Finished
None so far

**Comment by jrfnl** [comment]
@Mike-Hermans, @marconmartins, @Ipstenu, @NielsdeBlaauw, @fkeijzer, @GaryJones, @ckanitz Just wanted to drop you all a note that I'm sorry that I haven't managed to review your PRs yet, even though I'm so incredibly happy with them all.
Too many interesting conversations during this #WCEU kept me away from my computer.
I promise that I will get to them all within the week once I'm back home in a few days.

**Comment by jrfnl** [comment]
@marconmartins, @Ipstenu, @NielsdeBlaauw, @GaryJones, @ckanitz I'm planning to release WPCS 2.2.0 in the very near future and I'd love to include as many of the docs as possible in that release.

Aside from the PRs @NielsdeBlaauw send in today, are there any open docs PRs which have been updated after review, but haven't had attention since ?

If so: my apologies. Please do leave a comment either here or on the PR itself to draw attention to it, so we can try to get the PR merged before the release.

Also - if a PR does still need an update, but you either don't have time to do so yourself or have lost interest, please leave a comment on the PR and we'll try and have someone else finish it off.

I really appreciate the work you've done on this and would love to have it go out in the world and help people when they are working with WPCS.

**Comment by janw-me** [comment]
After about 2 hours of I figured out. How to make an exception for snake_case:

```
	<rule ref="WordPress.NamingConventions.ValidVariableName">
		<properties>
			<property name="customPropertiesWhitelist[]" value="hive_ID"/>
		</properties>
	</rule>
```

Now the question is where should this be documented.
I did found the [xml docs](https://github.com/WordPress/WordPress-Coding-Standards/blob/develop/WordPress/Docs/NamingConventions/ValidPostTypeSlugStandard.xml).
I don't think the format matches, I might be wrong.
Is there a place where this should be documented?
Can someone nudge me in the right direction?

**Comment by jrfnl** [comment]
@janw-me Sorry it took you so long. Next time, you may want to have a look at the wiki (where it is already documented): https://github.com/WordPress/WordPress-Coding-Standards/wiki/Customizable-sniff-properties#mixed-case-property-name-exceptions

**Comment by janw-me** [comment]
With googling I didn't find it...
And on the github pageI was looking for some "documentation" link but I should have looked for the wiki.
I probably should call it a day....

And of course I'll follow the wiki format. the `[]` is fugly.

**Comment by jrfnl** [comment]
@janw-me Well, the readme also contains a link to the wiki and mentions customizing sniff behaviour: https://github.com/WordPress/WordPress-Coding-Standards#customizing-sniff-behaviour
But if there are other ways in which you think this can be done which will help people find it more easily, I'd welcome a PR ;-)

**Comment by jrfnl** [comment]
@GaryJones @dingo-d 

* I've just updated the sniff list in the description to be in line with the current sniffs in WPCS.
* I'd like to start running XML lint over the docs and preferably also check them against an XSD.
    PHPCS does not provide an XSD and as there is discussion on what form documentation should take in the future I'm not sure pulling an XSD for the docs there would be PR which would be merged.
    So, what about creating an XSD and adding it to PHPCSUtils ? If the upstream PHPCS PR would be accepted, we can defer to that, but in the mean time we'd have something against which the docs can be automatically validated for the most basic of QA via a check in the CI.

Opinions ? Anyone up for creating the XSD ?

**Comment by dingo-d** [comment]
I went and played around with PhpStorm's automatic schema definitions and I got this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<xs:schema attributeFormDefault="unqualified" elementFormDefault="qualified" xmlns:xs="http://www.w3.org/2001/XMLSchema">
  <xs:element name="documentation" type="documentationType"/>
  <xs:complexType name="codeType">
    <xs:simpleContent>
      <xs:extension base="xs:string">
        <xs:attribute type="xs:string" name="title" use="optional"/>
      </xs:extension>
    </xs:simpleContent>
  </xs:complexType>
  <xs:complexType name="code_comparisonType">
    <xs:sequence>
      <xs:element type="codeType" name="code" maxOccurs="unbounded" minOccurs="0"/>
    </xs:sequence>
  </xs:complexType>
  <xs:complexType name="documentationType">
    <xs:sequence>
      <xs:element type="xs:string" name="standard"/>
      <xs:element type="code_comparisonType" name="code_comparison"/>
    </xs:sequence>
    <xs:attribute type="xs:string" name="title"/>
  </xs:complexType>
</xs:schema>
```

In one case (where I used `ArrayIndentationStandard.xml` as a basis for xsd generation) I got a difference in the `<xs:complexType name="documentationType">` part:

```xml
<xs:complexType name="documentationType">
  <xs:choice maxOccurs="unbounded" minOccurs="0">
    <xs:element type="xs:string" name="standard"/>
    <xs:element type="code_comparisonType" name="code_comparison"/>
  </xs:choice>
  <xs:attribute type="xs:string" name="title"/>
</xs:complexType>
```

The `xs:choice` vs `xs:sequence`.

I'm not all too familiar with how xds needs to be structured. Do we need to just add that part inside the `<xs:complexType name="documentationType">` tag?

**Comment by jrfnl** [comment]
@dingo-d Nice starting point, though it can probably be abstracted a little more.
An XSD basically is a definition of what is elements and attributes are allowed within an XML document, how they can be nested and if they are expected in a certain order etc etc.

The W3Schools information is not bad to get a general impression of the definition language: https://www.w3schools.com/xml/schema_intro.asp
Alternatively, for a deep dive, there are the official specs: https://www.w3.org/TR/xmlschema11-1/ ;-)

**Comment by jrfnl** [comment]
Oh and for inspiration, you could also have a look at the XSD for PHPCS rulesets: https://github.com/squizlabs/PHP_CodeSniffer/blob/master/phpcs.xsd

**Comment by rodrigoprimo** [comment]
> WordPress.PHP.TypeCasts - @anukasha-mo
> WordPress.WP.DiscouragedFunctions - @Nic-Sevic

@anukasha-mo and @Nic-Sevic, just checking if you are still planning on submitting a PR for the sniffs above, as it has been a few months since those two were claimed?

If you are not planning on working on the documentation for them, please let us know so that others can claim it.

**Comment by rodrigoprimo** [comment]
Just noting that I've updated the issue description ahead of WordCamp Asia 2026 Contributor Day. The main changes:

- Made the event and contact references generic instead of referencing WordCamp Europe 2019.
- Added a few clarifications to the documentation instructions (file naming format, XSD declaration, `<em>` placement, etc), based on patterns I noticed in previous doc PR reviews.
- Updated the list of sniffs that need documentation and created a new "Up for grabs (previously claimed)" section.
- Added links to a few WPCS sniff docs as additional examples to follow instead of linking just to PHPCS sniff docs.

**Comment by thisismyurl** [comment]
Before I open a PR, I'd like to check the scope of one of the contributing guidelines.

I have documentation ready for `WordPress.WP.DiscouragedFunctions`, one of the four sniffs still on the To Do list. I should be upfront about how it was produced: I read the sniff source and its unit test fixtures to work out what it actually checks, then used an AI assistant to help draft the prose, and verified the result by rendering it with the text generator and validating it against `phpcsdocs.xsd`.

`CONTRIBUTING.md` says not to submit AI-generated code. I read that as written and stopped rather than assume it only applied to PHP. Does it cover sniff documentation prose as well? I also see #2735 is currently revising these guidelines, so I may be asking mid-change.

If documentation is covered, that is completely fine. I will either rewrite it from scratch in my own words or leave the sniff for someone else, and I will not open the PR. I would rather ask than spend your review time on something you have already said you do not want.

**Comment by jrfnl** [comment]
@thisismyurl I very much appreciate you asking. At this time, the AI-ban is for everything, including documentation. Whether the open PR will change that remains to be seen.

Speaking for me personally, my eyes glace over after reading two sentences of AI generated text and I just stop reading. I don't think that's the effect we'd like the WPCS documentation to have.

> I read the sniff source and its unit test fixtures to work out what it actually checks

To be honest, if you've already done that, then writing the docs manually would probably have been quicker than writing the AI prompt....

**Comment by thisismyurl** [comment]
Understood, and thank you for the clear answer. I will not open the PR.

Fair point on the last bit as well. I am leaving `WordPress.WP.DiscouragedFunctions` unclaimed on the To Do list so it stays available for whoever gets to it, myself included if I come back and write it properly.
